### ISSUE TITLE
Rule feature request: Using derivedStateOf without any State object in calculation

### ISSUE THREAD

**Issue by svenjacobs** [twitter/compose-rules#88]
Currently there is no (lint) warning when using `derivedStateOf` without accessing any `State` object in its calculation, which can lead to unexpected behaviour. Since no `State` object is read, the derived value will never update which however might not be obvious just by looking at the code.

Let's take the following Composable for example:

```kotlin
@Composable
fun Amount(
    amount: Int,
    modifier: Modifier = Modifier,
) {
    val isZeroAmount by remember { derivedStateOf { amount == 0 } }

    Box(modifier = modifier) {
        if (isZeroAmount) {
            Text("Amount: Is zero")
        } else {
            Text("Amount: $amount")
        }
    }
}
```

If `amount` is `1` initially and becomes `0` on the next recomposition, the Composable will incorrectly display `Amount: 0` instead of `Amount: Is zero`.

This might for instance happen when a `State` variable with `by` delegation is changed to a plain type.
The correct code for this case would be:

```kotlin
@Composable
fun Amount(
    amount: State<Int>,
    modifier: Modifier = Modifier,
) {
    val isZeroAmount by remember { derivedStateOf { amount.value == 0 } }

    Box(modifier = modifier) {
        if (isZeroAmount) {
            Text("Amount: Is zero")
        } else {
            Text("Amount: ${amount.value}")
        }
    }
}
```

**Comment by mrmans0n** [comment]
Interesting idea! 

Doing this w/o type resolution might prove hard but I think it might be doable with some heuristics.

**Comment by svenjacobs** [comment]
I assumed that this will be a tricky one without type resolution in the context of the rule's code. Also just ensuring that any read inside the calculation must be done via `state.value` is not sufficient when `by` delegation is used.

**Comment by chrisbanes** [comment]
I like it! It's going to be hard to determine whether something is backed by state though. There's the obvious case of having a `State` object in the same function, but the state could be anywhere in the call stack, and possible hidden through lambdas or other objects.

I think a better way is for `derivedStateOf` to throw an exception, when it detects that nothing state backed has been read.

**Comment by svenjacobs** [comment]
I like the idea of `derivedStateOf` throwing an exception. It even becomes trickier when a `State` is read inside the calculation which however is not part of the result value. Then the derived state only updates when the unrelated state is updated, which might be even more confusing.

```
val isZeroAmount by remember { 
    derivedStateOf {
        log("Product type is: $productType") // productType is State<String> for example
        amount == 1
    }
```

I think only the Compose compiler or runtime can detect those cases for certain.
